R/tree.R

```diff
@@ -0,0 +1,314 @@
+    x_attributes <- x_attributes[names(x_attributes) != "names"]
+  }
+  has_attributes <- length(x_attributes) > 0 & opts$show_attributes
+
+  # ===== Start recursion logic
+  # Turn into a s3 method for recursion
+  if(!is.atomic(x) & depth <= opts$max_depth & !is.function(x) & !is.environment(x)){
+    children <- as.list(x)
+
+    # Traverse children, if any exist
+    n_children <- length(children)
+
+    # If children have names, give them the names
+    for (i in seq_along(children)) {
+      id <- names(x)[i]
+      if(is.null(id) & opts$index_arraylike) id <- i
+
+      child_type <- if(i < n_children){
+        "child"
+      } else if(has_attributes) {
+        "pre-attrs"
+      } else {
+        "last-child"
+      }
+      tree_internal(
+        x = children[[i]],
+        x_id = id,
+        branch_hist = c(branch_hist, child_type),
+        opts = opts
+      )
+    }
+  }
+  # ===== End recursion logic
+
+  # Add any attributes as an "attr" prefixed children at end
+  if(has_attributes){
+    n_attributes <- length(x_attributes)
+    for(i in seq_len(n_attributes)){
+      tree_internal(
+        x = x_attributes[[i]],
+        x_id = crayon::italic(paste0("<attr>", names(x_attributes)[i])),
+        opts = opts,
+        branch_hist = c(branch_hist, paste0(if(i == n_attributes) "last-", "attribute")),
+        attr_mode = TRUE # Let tree know this is an attribute
+      )
+    }
+  }
+}
+
+#' Build element or node label in tree
+#'
+#' These methods control how the value of a given node is printed. New methods
+#' can be added if support is needed for a novel class
+#'
+#' @inheritParams tree
+#'
+#' @export
+tree_label <- function(x, val_printer, class_printer, remove_newlines){
+  UseMethod("tree_label")
+}
+
+#' @export
+tree_label.function <- function(x, ...){
+  crayon::italic("function(){...}")
+}
+
+#' @export
+tree_label.environment <- function(x,...){
+  format.default(x)
+}
+
+#' @export
+tree_label.NULL <- function(x,...){
+  "<NULL>"
+}
+
+#' @export
+tree_label.character <- function(x, remove_newlines, ...){
+
+  # Get rid of new-line so they don't break tree flow
+  if(remove_newlines){
+    x <- gsub("\\n", replacement = " ", x = x, perl = TRUE)
+  }
+
+  # Shorten strings if needed
+  max_standalone_length <- 35
+  max_vec_length <- 15
+  max_length <- if(length(x) == 1) max_standalone_length else max_vec_length
+  x <- ifelse(
+    nchar(x) > max_length,
+    # Since we add an elipses we need to take a bit more than the max length
+    # off. The gsub adds elipses but also makes sure we dont awkwardly end on
+    # a space.
+    gsub(x = substr(x, start = 1, max_length - 3),
+         pattern = "\\s*$",
+         replacement = "...",
+         perl = TRUE),
+    x
+  )
+
+  tree_label.default(paste0("\"", x, "\""),...)
+}
+
+
+
+#' @export
+tree_label.default <- function(x, val_printer, class_printer, remove_newlines){
+
+  # There are a few psuedo-types that we want different printing behavior for.
```

Comment:
I don't think you really need these comments here, because I think you can assume the reader of lobstr source code will already be familiar with these issues.